### ISSUE TITLE
Represent the state of the chaise/search app in the URL

### ISSUE THREAD

**Issue by robes** [informatics-isi-edu/chaise#1]
A Chaise URL should represent much (if not all) of the state of the application. For example, the state of the app currently includes things like:
* the current collection (entity set) that the user is browsing,
* the facets that the user has selected,
* the current layout (list, card, or table),
* whether the slide out is shown and which panel is displayed,
* and if we had sort ordering that would also be part of the state.

All of this state could and should be represented in the URL. If it were represented in the URL then a user could:
* link directly to it from another web page,
* bookmark it for later viewing, or
* share it (email, text, etc) to a colleague for them to see the exact same view.

To do this properly, the state of the application should come in the fragment identifier portion of the URL scheme:
http://en.wikipedia.org/wiki/Fragment_identifier

That means that it should be after the "#" character in the URL like:
https://example.org/myapp?someoptions#fragment-id-here

**Comment by svoinea** [comment]
The current bookmark contains the state including the selected collection as well as the selected facets.

**Comment by svoinea** [comment]
Added in the current bookmark the current layout as part of the state of the application.

**Comment by robes** [comment]
we should all review the status of this and the design of the URL itself

**Comment by robes** [comment]
@carlkesselman: @karlcz, @svoinea and I met this afternoon to discuss the current state of this issue. We sketched out a fragment id for this feature as:

```
# <route> ? catalog = <id> & entity = <schema>:<table> & facets = ( <attribute> <op> ( <value> | ( <value>, <value>, ... ), ... ) & other_tbd &...
```
where:
- `route` : according to @svoinea is required by AngularJS in the fragment id for our configuration. We can pick the name so we have a chance of making this meaningful like `...#/search?...`.
- `id`: the catalog id
- `schema:table`: the qualified table name that is the target of the faceting. __TBD__: usage of `:`
- `attribute`: the qualified `schema:table:column` name used in the faceting predicate
- `op`: the operator in the faceting predicate, which may require either URL encoding of `=`, `>`, etc. or our own more user-friendly encoding such as `:eq:`, `:gt:`, etc.
- `value`: the value in the faceting predicate
- `other_tbd`: a placeholder for one or more parameters needed to represent the state of the UI beyond the basic parameters needed for faceting. These may include the current layout (list, card, table) etc.

At a minimum, all parameter values (id, schema, table, attribute, value, etc.) would need to be URL encoded. But as long as no special characters are used in them (such as spaces) the value will be unchanged and user-friendly.

In the `facets` parameter, it should support a conjunction of disjunctions. I.e., `facets = ( a1 :eq: val1, a2 :eq: ( val20, val21, val23 ) )` is interpreted as facet _a1_ equals _val1_ __and__ _a2_ equals _val20_ or _val21_ or _val23_.

@svoinea looked into the allowable characters in the fragment (as allowed by AngularJS). He has a preliminary results of that. I think it would be good to get a demonstration tomorrow before finalizing on the format of the fragment identifier.

**Comment by karlcz** [comment]
We cannot use the <attribute> <op> <values> syntax if <op> will be
url-quoting '='.  There would be no way to distinguish an attribute
that has an '=' encoded in it from an '=' operator.

All user-generated values (schema, table, column, value) need to
support url-quoting. The meta-syntax that separates user-generated
values must begin and end with non-quoted "reserved" characters in
order to be effective boundary markers for tokenization.  It is only
these "reserved" characters that can have a different meaning whether
quoted or unquoted according to RFC rules.

**Comment by mikedarcy** [comment]
We also need a way for the app to expose only the ermrest query (or set of queries) that constitute the current facet-filtered collection.  This probably needs a separate issue opened for it, but since it is closely related implementation-wise, I wanted to raise the point here.

**Comment by karlcz** [comment]
This should also include pagination state so we can have links to any part of a large result set. The pagination would include something like the first (or last) row key to display and the number of rows.

**Comment by svoinea** [comment]
The fragment part of the URL is automatically decoded URI.
That means that the bookmark can not rely on the encoded values it has.
Therefore, in parsing the bookmark, we can not make any more a distinction between the tokens delimiters and the values containing such "encoded" URI delimiters.

**Comment by robes** [comment]
I wrote a very simple test page that seems to contradict the above statement. And @karlcz provided a URL that tests it. This appears to show that the `window.location` can be used to get the current and properly (?) encoded URL string.

```
http://www.isi.edu/~schuler/test.html#%5C%21%2A%27%28%29%3B%3A%40%26%3D%2B%24%2C%2F%3F%23%25!*();:@&=+$,/?
```

You might be able to [click here](http://www.isi.edu/~schuler/test.html#%5C%21%2A%27%28%29%3B%3A%40%26%3D%2B%24%2C%2F%3F%23%25!*();:@&=+$,/?) to follow that test URL.

But be careful because there are many inconsistencies when copying URLs to and from clipboards, browser location bars, and github enhanced markdown (!) to name a few.

Also, do not trust what you see in the browser location bar as a reflection of the current state of `window.location`. Do not trust the string that gets printed in the html page. We should hopefully be able to trust the output printed using `console.log(...)` and we certainly better be able to trust the display by the browser's debug tools.

Unless someone shows this to be wrong, it appears that we should be able to move forward with the URL scheme as specified.

**Comment by svoinea** [comment]
The simple test page works, but not in the Angular environment.
Once you are in an Angular controller, the URL fragment is already decoded.
You might catch the original URL before entering Angular, but this is possible only when you load a JavaScript file, and this usually happens only once, when you launch the application.

**Comment by karlcz** [comment]
You're saying that Angular rewrites the location.href attribute, corrupting it?

**Comment by robes** [comment]
Please point us to the test code used to make this conclusion.

**Comment by svoinea** [comment]
It re-writes the location.hash attribute by applying the decodeURIComponent() function.  
The location.hash is the URL fragment, therefore it is part of the location.href. 

I have defined in ermrest.js (not an Angular controller) the variable:

var WINDOW_LOCATION = window.location.href;

When the application is launched,  in the Angular controller I'm tracing the WINDOW_LOCATION variable and the value of window.location.href and I can see the difference.

Each time I enter a new URL, the trace for window.location.href shows the new value (decoded), while WINDOW_LOCATION is unchanged as the ermrest.js file was not reloaded.
 
It is possible that this behaviour is related to Angular usage of the URL fragment for routing.

**Comment by karlcz** [comment]
When you say you "entered a new URL" what do you mean specifically?

- Visited a new link?
- Pasted into browser?
- Assigned to location?
- Used some other API?

**Comment by svoinea** [comment]
I copied a value by right clicking the Bookmark. Its value is:

https://dev.facebase.org/ermrest-ui/app.html#/search?entity=legacy:dataset&facets=(dataset:summary::ciregexp::location%3B)&layout=list&page=1

I pasted this value in the URL browser. The trace for window.location.href shows:

https://dev.facebase.org/ermrest-ui/app.html#/search?entity=legacy:dataset&facets=(dataset:summary::ciregexp::location;)&layout=list&page=1

Therefore, %3B was decoded to ";".

**Comment by robes** [comment]
Copy and paste is problematic with browser location bars, as we are learning. A lot of things could go wrong and are beyond our control. However, I did copy your last URL into my little Angular test app like this.

http://www.isi.edu/~schuler/angular-test/#/search?entity=legacy:dataset&facets=(dataset:summary::ciregexp::location%3B)&layout=list&page=1

The %3B appears to be preserved in the window.location.href variable.

**Comment by robes** [comment]
Okay, so I looked in the debugger at `window.location.hash` versus `window.location.href` and the `hash` does indeed appear to be decoded whereas the `href` does not. I then found this little blog entry on the web that says that Firefox has a bug in it where it decodes the character in the `hash` and basically says “don’t use it” just use `href` instead and split it.

https://www.quora.com/Does-window-location-hash-work-across-all-browsers

On September 15, 2015 at 11:07:28 AM, svoinea (notifications@github.com) wrote:

It re-writes the location.hash attribute by applying the decodeURIComponent() function.

The location.hash is the URL fragment, therefore it is part of the location.href.

I have defined in ermrest.js (not an Angular controller) the variable:

var WINDOW_LOCATION = window.location.href;

When the application is launched, in the Angular controller I'm tracing the WINDOW_LOCATION variable and the value of window.location.href and I can see the difference.

Each time I enter a new URL, the trace for window.location.href shows the new value (decoded), while WINDOW_LOCATION is unchanged as the ermrest.js file was not reloaded.

It is possible that this behaviour is related to Angular usage of the URL fragment for routing.

**Comment by svoinea** [comment]
Just to clarify that in my test I don't copy from the browser URL box and I'm referring to the href value and not to the hash one.

A link that might be useful: https://docs.angularjs.org/guide/$location

It specifies:

When you call the absUrl() method, the returned value is a full url with its segments encoded.

However, my test does not prove this statement.

**Comment by svoinea** [comment]
I have "extended" the little Angular test app to work in a view router mode:

https://serbancentos.isi.edu/angular-test#/search?entity=legacy:dataset&facets=(dataset:summary::ciregexp::location%3B,dataset_data_type:data_type::eq::(Enhancer%20reporter%20data;OPT%20images),dataset_chromosome:start_position::geq::(57096119;4269384),dataset_chromosome:start_position::leq::(163330104;194677837))&layout=list&page=1

The URL fragment is decoded in the same manner as in chaise:

URL: https://serbancentos.isi.edu/angular-test/#/search?entity=legacy:dataset&facets=(dataset:summary::ciregexp::location;,dataset_data_type:data_type::eq::(Enhancer%20reporter%20data;OPT%20images),dataset_chromosome:start_position::geq::(57096119;4269384),dataset_chromosome:start_position::leq::(163330104;194677837))&layout=list&page=1

**Comment by robes** [comment]
I updated the angular test app too. The `$location` service encodes and decodes values corresponding to the window location. Essentially stated here in the Angular docs:

https://docs.angularjs.org/api/ng/service/$location
https://docs.angularjs.org/guide/$location

I looked workarounds for this automatic encoding/decoding but haven't found any. I am not sure whether we can live without the services provided by `$location` since it wraps the HTML5 history API and provides backward compatibility with older browsers.

The `$route` service (or `ngRoute`) depends on `$location` so therefore it will inherit its behavior.

https://docs.angularjs.org/api/ngRoute/service/$route
